### PR TITLE
[libc] Extend the baremetal I/O vendor ABI

### PR DESCRIPTION
**Pull request by petrhosek**
This refines and extends the external ABI for I/O, later changes will update the baremetal implementations of I/O functions to use these.

### CODE REVIEW

File: libc/src/__support/OSUtil/baremetal/io.cpp
Comment (frobtech) [inline libc/src/__support/OSUtil/baremetal/io.cpp:49]: The cast is superfluous but does no harm.

File: libc/src/__support/OSUtil/baremetal/io.cpp
Comment (frobtech) [inline libc/src/__support/OSUtil/baremetal/io.cpp:40]: I think the comments here need to be a bit more clear that these three symbols must be provided, even if they don't point at anything, and give examples of how to meet the ABI.

All these declarations belong in some header file. Eventually such a header file should be installed somewhere so embedders can use it for the ABI contract  they're meeting.  But that can be a later cleanup, a comment here is enough for now.

File: libc/src/__support/OSUtil/baremetal/io.cpp
Comment (frobtech) [inline libc/src/__support/OSUtil/baremetal/io.cpp:53]: I think we should change these to propagate return values, and leave it to the embedder to decide if they want to return apparent success in error (or bit-bucket) cases.
But that can be a follow-on cleanup after this change lands.

**Review by frobtech — APPROVED** [review]
This is a good incremental step and we can continue the cleanup and formalization with later changes.

**Review by michaelrj-google — COMMENTED** [review]
LGTM, I agree with the comments. For return types I think we should generally match what `fread` and `fwrite` do: return the number of characters read/written.

File: libc/src/__support/OSUtil/baremetal/io.cpp
Comment (petrhosek) [reply libc/src/__support/OSUtil/baremetal/io.cpp:53]: I agree, but I'd like to do that in a follow-up change since I'll also need to update the callers.

**Comment by petrhosek** [comment]
> LGTM, I agree with the comments. For return types I think we should generally match what `fread` and `fwrite` do: return the number of characters read/written.

I plan to change the signatures and update all callers in a follow-up change.

**Comment by frobtech** [comment]
> LGTM, I agree with the comments. For return types I think we should generally match what `fread` and `fwrite` do: return the number of characters read/written.

The precise protocol here is already set by the `fopencookie` callbacks' APIs, which have the same return value semantics as their POSIX counterparts.

**Pull request merged**
